### PR TITLE
SNOW-951722: Fix offset impacting previous limit

### PR DESCRIPTION
**Pull request by sfc-gh-yixie**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-951722

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: tests/integ/test_simplifier_suite.py [hunk 1]
```diff
@@ -747,6 +747,45 @@ def test_filter_order_limit_together(session, simplifier_table):
...
+    assert (
+        df4.queries["queries"][-1]
+        == f"SELECT  *  FROM (select * from {simplifier_table}) LIMIT 10"
+    )
+
+    df5 = df4.limit(5, offset=2)
+    assert (
+        df5.queries["queries"][-1]
+        == f"SELECT  *  FROM ( SELECT  *  FROM (select * from {simplifier_table}) LIMIT 10) "
+        f"LIMIT 5 OFFSET 2"
```
Comment (sfc-gh-stan) [inline tests/integ/test_simplifier_suite.py:785]: Why can't we flatten the `LIMIT 10` and the `LIMIT 5 OFFSET 2` into `LIMIT 5 OFFSET 2`?

Comment (sfc-gh-yixie) [reply tests/integ/test_simplifier_suite.py:785]: My test may have a bad example. But `LIMIT 10` and `LIMIT 5 OFFSET 20` can't be flattened to `LIMIT 5 OFFSET 20`.
This is a corner case though.

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 2]
```diff
@@ -766,11 +766,19 @@ def set_operator(
         return new
 
     def limit(self, n: int, *, offset: int = 0) -> "SelectStatement":
-        new = copy(self)
-        new.from_ = self.from_.to_subqueryable()
-        new.limit_ = min(self.limit_, n) if self.limit_ else n
-        new.offset = (self.offset + offset) if self.offset else offset
-        new.column_states = self.column_states
+        if offset and self.limit_:  # Don't flatten when there both layers have offset.
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:769]: The documentation is suggesting `if offset and self.offset`? Shall we modify the documentation?

File: tests/integ/test_simplifier_suite.py (hunk 1 above)
Comment (sfc-gh-stan) [reply tests/integ/test_simplifier_suite.py:785]: Yea, I wonder if this edge case could be optimized (to allow more flattening) by doing some extra check, but maybe this is unnecessary. What do you think if we create a ticket for potential optimization?

Comment (sfc-gh-yixie) [reply tests/integ/test_simplifier_suite.py]: Updated the test to use offset 20 so easier to understand.

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py (hunk 2 above)
Comment (sfc-gh-yixie) [reply src/snowflake/snowpark/_internal/analyzer/select_statement.py]: Good catch. Updated the comments.

**Review by sfc-gh-mabrennan — APPROVED** [review]
lgtm

File: tests/integ/test_simplifier_suite.py (hunk 1 above)
Comment (sfc-gh-yixie) [reply tests/integ/test_simplifier_suite.py]: @sfc-gh-stan Created a JIRA to follow-up anyways to see if we can use more detailed rules to further flatten.

**Review by sfc-gh-stan — APPROVED** [review]

**Review by sfc-gh-jdu — APPROVED** [review]
